r-data-manipulation.Rmd

```diff
@@ -0,0 +1,1111 @@
+
+A call to `ifelse()` takes the form:
+```r
+ifelse(test, yes, no)
+```
+Where `test` is a logical expression, `yes` the value for the elements that return `TRUE`, and `no` the value for the elements that return `FALSE`. (Both `yes` and `no` could be other column names, in which case, the corresponding element of `yes` would be returned for `TRUE` elements).
+```{r}
+dog_licenses %>% 
+  mutate(
+    called_chase = ifelse(animal_name == "CHASE", "called_chase", "not_called_chase")
+    ) %>% 
+  select(animal_name, called_chase)
+```
+
+You can perform multiple mutate steps at once by passing more arguments to `mutate()`, so an alternative way of writing the above code (with more keystrokes, but with lines that are shorter) would be:
+```{r}
+dog_licenses %>% 
+  mutate(
+    is_chase = animal_name == "CHASE",
+    called_chase = ifelse(is_chase, "called_chase", "not_called_chase")
+    ) %>% 
+  select(animal_name, is_chase, called_chase)
+```
+
+Arithmetic is another common operation that returns as many elements as there are rows.  For instance we could see how long licenses are issued for:
+```{r}
+dog_licenses %>% 
+  mutate(license_duration = license_expired_date - license_issued_date) %>% 
+  select(license_duration)
+```
+Interesting that there is so much variation here. 
+
+TODO: provide link to RStudio data mini cheatsheet with other functions that are useful with mutate.
+
+You can use `mutate()` with operations that give one number based on all the rows:
+```{r}
+dog_licenses %>% 
+  mutate(
+    license_duration = license_expired_date - license_issued_date,
+    avg_duration = mean(license_duration)) %>% 
+  select(license_duration, avg_duration)
+```
+You'll get back the original number of rows, but the single value will be repeated in all of them - on average licenses are issued for 467.321 days.  You'll see a different manipulation, `summarise()` that collapses many rows into one later in this chapter.
+
+### Exercise: Ages of dogs
+
+Use `mutate()` to add a column `age_at_issue` that contains the dogs approximate age on the day the license was issued.
+
+### Exercise: Mutts
+
+Use `mutate()` along with `ifelse()` to create a column `breed` that takes values `"unknown"` if `breed_name` is `"Unknown"` and `"known"` otherwise.
+
+*Extra challenge* Can you figure out if the licenses issued to unknown breed dogs are of longer or shorter duration on average than known breed dogs?
+
+### Exercise: Name length
+
+The function `str_length()` in the stringr package finds the length of character strings.  Replace the `___` in following code to add a column called `name_length` that contains the length of the dog's name:
+```{r, eval = FALSE}
+dog_licenses %>% 
+  ___(___ = stringr::str_length(animal_name))
+```
+
+Now add an `arrange()` step to find the licenses issued to dogs with the longest names?
+
+## How can I tell what's gone wrong in my programs?
```

Comment:
I agree it feels a little out of place.  

It does seem like a discussion that should happen before too much is covered, but it doesn't have to be here. If I was teaching in a semester course I feel like I'd cover this at the end of lecture 1 or start of lecture 2...maybe not in its entirety, but at least to say "things will go wrong, and that's OK, here's some tips on getting things going again".

Its position in the book probably doesn't matter to much for an instructor-led class since they can jump to it when they feel it's appropriate (as long as the examples don't assume too much).  Where would it be most useful, and least distracting, for someone working through the book on their own?

Reply:
This will require some re-writing if we move it earlier to use only functions learners have seen.  Leaving as-is for this PR, and creating an issue to re-visit. See  #188